### ISSUE TITLE
The IncorrectCase diagnostic is overzealous for types that should be Camel Cased

### ISSUE THREAD

**Issue by ArifRoktim** [rust-analyzer/rust-analyzer#6305]
For the following `lib.rs`:
```rust
pub struct AxisAlignedBoundingBox; // A
pub struct AABB; // B
pub struct A_A_B_B; // C
```
`cargo check` will only mark version "C" as having incorrect case:
```sh
$ cargo check
warning: type `A_A_B_B` should have an upper camel case name
 --> src/lib.rs:3:12
  |
3 | pub struct A_A_B_B; // C
  |            ^^^^^^^ help: convert the identifier to upper camel case: `AABB`
```

While rust-analyzer will erroneously mark version "B" as having incorrect case, saying 
> Hint [rust-analyzer] Structure `AABB` should have CamelCase name, e.g. `Aabb`

I agree with cargo here. I'd say `AABB` isn't SCREAMING_SNAKE_CASE since there are no underscores.

If you guys agree that this is a bug, I'd like to open a PR for this.

**Comment by matklad** [comment]
cc @popzxc

**Comment by matklad** [comment]
Yes, this is a bug, a fix would be welcome!

Though, as per [API guildelines](https://rust-lang.github.io/api-guidelines/naming.html), the name `Aabb` would be better.

**Comment by popzxc** [comment]
Well, personally I would have considered `AABB` a screaming case still, but I agree that diagnostic must match the compiler logic indeed.

**Comment by ArifRoktim** [comment]
Doing some quick research, it appears that the rules regarding acronyms in camel cased symbols are unclear. I'm now leaning more towards `Aabb` over `AABB` than before.
I'll still open that PR so that rust-analyzer matches what the compiler does.

**Comment by iago-lito** [comment]
FWIW, I've just ran into the following:

![SHIFTS](https://user-images.githubusercontent.com/10091387/97222058-0f295180-17ce-11eb-96bb-9d1938a1d123.png)

Although `cargo check` yields no warning. Is that related?
